### PR TITLE
HEEDLS-652 Add back end code to add resource to action plan

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-652

### Description
Add back end code for the Add to action plan functionality

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [ ] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs:64]: Do you think it would be worth checking the other columns are null, just to make sure illogical additions are not made (e.g. CompletedDate not being null etc.)?

File: DigitalLearningSolutions.Data.Tests/TestHelpers/LearningLogItemsTestHelper.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/TestHelpers/LearningLogItemsTestHelper.cs:54]: We are going to be using LearningLogItems in a bunch of other tickets. Do you reckon its worth moving this class into its own file under Models, to reduce the number of potentially redundant classes when everything gets merged?

File: DigitalLearningSolutions.Data.Tests/Services/ActionPlanServiceTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/Services/ActionPlanServiceTests.cs:107]: At the risk of making this test even larger, what do you think of additionally adding a similar set of assertions for this to test the competency Ids are the correct ones, such as (excuse the formatting):

```
 var expectedMatchingCompetencies = new[] { 2, 3, 5, 6, 8 };
            foreach (var competencyId in expectedMatchingCompetencies)
            {
                A.CallTo(
                    () => learningLogItemsDataService.InsertLearningLogItemCompetencies(learningLogId, competencyId, addedDate)
                ).MustHaveHappenedOnceExactly(); 
            }
```
Would probably need expectedMatchingCompetencies to be defined alongside the other collections of competencies somewhere obvious.

**Review by DanBloxham-sw — COMMENTED** [review]
Looks good on the whole. Left a few comments about potential improvements.

Unfortunately, at time of writing Kevin has not confirmed my guess at the LearningLogItemCompetencies insertion process, but I feel pretty confident with what we have here. With that, and the state of the API client work, feel free to argue against the suggestions in the comments.

**Comment by DanBloxham-sw** [comment]
Kevin has just confirmed that the process for adding things to LearningLogItemCompetencies is exactly how we have implemented it here, so no need to worry about changing it.

File: DigitalLearningSolutions.Data.Tests/DataServices/CompetencyLearningResourcesDataServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/CompetencyLearningResourcesDataServiceTests.cs:82]: I don't even need to do that, just don't assign them to variables.

File: DigitalLearningSolutions.Data.Tests/DataServices/CompetencyLearningResourcesDataServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/CompetencyLearningResourcesDataServiceTests.cs:74]: I don't see any sensible way to do it without identity insert since we won't be able to get the item by Id since the Id will auto-increment.

Given it is against the test database though, I don't think it'll matter as long as we assume we always insert records in transactions like we are doing here (i.e. no permanent test data ever) or we create some permanent test data right now.

Otherwise I think the only options that would be more future proof would be to insert them at the max int value (unlikely to hit this) or delete all data in the transaction (which would likely violate some foreign keys if we did have data)

File: DigitalLearningSolutions.Data.Tests/TestHelpers/CompetencyLearningResourcesTestHelper.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/TestHelpers/CompetencyLearningResourcesTestHelper.cs:28]: Copy and paste error, will fix.

File: DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs:64]: Yeah probably.
If I'm adding a new LearningLogItems class for the below, I guess it needs to have properties for all the columns in the database, and we can then check the rest are null with a full select

File: DigitalLearningSolutions.Data.Tests/TestHelpers/LearningLogItemsTestHelper.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/TestHelpers/LearningLogItemsTestHelper.cs:54]: Added this with columns names as per the database. We'll probably want to rename columns as we actually use this properly later.

File: DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs:65]: I feel like the ActivityTypeId should be 8 explicitly, as it comes from a reference table, but I suppose we can't guarantee that given the migration. Do you think this is worth changing? An alternative might be making the query used JOIN onto the ActivityTypes table, and recovering the TypeLabel and making sure that is "Learning Hub Resource"

**Review by DanBloxham-sw — COMMENTED** [review]
I think this is good. I've left one comment about making a test more specific, not sure if it's worth it.

Also unsure if I should approve this given it doesn't use the real API client yet, but what I've seen so far is worthy of approval.

File: DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/LearningLogItemsDataServiceTests.cs]: Changed this to get the TypeLabel as I think it is safer than assuming an ID

**Review by DanBloxham-sw — COMMENTED** [review]
Yeah I think this is now good sans API client.

I'll not "approve" it yet, as I will wait until we're fully happy to merge with the API client, but I'm happy for this to be moved on.

**Review by DanBloxham-sw — APPROVED** [review]
Actually, I'm just approving this properly, so it can be merged as is, then a second MR can be made to incorporate the API client methods.

**Review by livzorn — APPROVED** [review]
Looks good!

**Pull request merged**
